app/post_upgrade.go

```diff
@@ -50,34 +60,57 @@ func PostUpgradeCmd() cli.Command {
 
 func postUpgrade(c *cli.Context) error {
 	namespace := c.String(FlagNamespace)
-	if namespace == "" {
-		return errors.New("namespace is required")
-	}
 
 	config, err := clientcmd.BuildConfigFromFlags("", c.String(FlagKubeConfig))
 	if err != nil {
 		return errors.Wrap(err, "failed to get client config")
 	}
 
+	eventBroadcaster, err := createEventBroadcaster(config)
+	if err != nil {
+		return errors.Wrap(err, "failed to create event broadcaster")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhorn.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return errors.Wrap(err, "failed to create scheme")
+	}
+
+	eventRecorder := eventBroadcaster.NewRecorder(scheme, corev1.EventSource{Component: PostUpgradeEventer})
+
 	kubeClient, err := kubernetes.NewForConfig(config)
 	if err != nil {
 		return errors.Wrap(err, "failed to get k8s client")
 	}
 
-	return newPostUpgrader(namespace, kubeClient).Run()
+	err = newPostUpgrader(namespace, kubeClient, eventRecorder).Run()
+
+	return err
 }
 
 type postUpgrader struct {
-	namespace  string
-	kubeClient kubernetes.Interface
+	namespace     string
+	kubeClient    kubernetes.Interface
+	eventRecorder record.EventRecorder
 }
 
-func newPostUpgrader(namespace string, kubeClient kubernetes.Interface) *postUpgrader {
-	return &postUpgrader{namespace, kubeClient}
+func newPostUpgrader(namespace string, kubeClient kubernetes.Interface, eventRecorder record.EventRecorder) *postUpgrader {
+	return &postUpgrader{namespace, kubeClient, eventRecorder}
 }
 
 func (u *postUpgrader) Run() error {
-	if err := u.waitManagerUpgradeComplete(); err != nil {
+	var err error
+	defer func() {
+		if err != nil {
+			u.eventRecorder.Event(&corev1.ObjectReference{Namespace: u.namespace, Name: PostUpgradeEventer},
+				corev1.EventTypeWarning, constant.EventReasonFailedUpgradePostCheck, err.Error())
+		} else {
+			u.eventRecorder.Event(&corev1.ObjectReference{Namespace: u.namespace, Name: PostUpgradeEventer},
+				corev1.EventTypeNormal, constant.EventReasonPassedUpgradeCheck, "post-upgrade check passed")
+		}
+	}()
+
+	if err = u.waitManagerUpgradeComplete(); err != nil {
```

Comment:
This is nonsense.  If modified as suggested, it won't compile because the anonymous defer function references `err`, which is undeclared.
If the declaration is kept, but the other change is made, that `err` will shadow the first one, and the defer will event a success, even though `waitManagerUpgradeComplete` failed.